Ensemble!

jmccreight opened the pull request:
All new add, compose, and run methods. 
No scheduler work yet. 
All tested!

jmills-ncar commented:
great coverage, nice job!

jmills-ncar commented on wrfhydropy/core/ensemble.py:
```diff
@@ -0,0 +1,369 @@
+
+        # TODO(JLM): Could this be parallelized?
+        all_diff_keys = set({})
+        for ii in range(1, len(self)):
```
I belive `enumerate()` is preferred over `range()`

jmills-ncar commented on wrfhydropy/core/ensemble.py:
```diff
@@ -0,0 +1,369 @@
+            diff_keys = list(diff['values_changed'].keys())
+            all_diff_keys = all_diff_keys | set([ss.replace('root', '') for ss in diff_keys])
+
+        # TODO(JLM): What is this doing? Comment.
```
haha, didn't you write it?

jmills-ncar commented on wrfhydropy/core/ensemble.py:
```diff
@@ -0,0 +1,369 @@
+        att_tuple: tuple,
+        values: list
+    ):
+
```
docstring please

jmills-ncar commented on wrfhydropy/core/ensemble.py:
```diff
@@ -0,0 +1,369 @@
+        self,
+        n_concurrent: int=1
+    ):
+
```
docstring please

jmills-ncar commented on wrfhydropy/core/ensemble_tools.py:
```diff
@@ -0,0 +1,84 @@
+
+
+def is_sub_obj(obj):
+    """Test if an object is has a __dict__ (may not be the best definition of an object, 
```
include args in docstring

jmccreight commented:
TRUE DAT

jmccreight commented on wrfhydropy/core/ensemble.py:
```diff
@@ -0,0 +1,369 @@
+
+
+def parallel_compose_addjobs(arg_dict):
+    for jj in arg_dict['jobs']:
```
I think we could debate this. I like double letters for iterands where the letters reference the thing being iterated. It makes clear what in a loop is changing.

jmccreight commented on wrfhydropy/core/ensemble.py:
```diff
@@ -0,0 +1,369 @@
+
+
+def parallel_compose_addscheduler(arg_dict):
+
```
OK. My pep8 dosent complain about this...

jmccreight commented on wrfhydropy/core/ensemble.py:
```diff
@@ -0,0 +1,369 @@
+        self,
+        obj: Union[list, Scheduler, Job]
+    ):
+        """Add an approparite object to an EnsembleSimulation, such as a Simulation, Job, or
```
sorry, stole this from simulation! :) Will update both

jmccreight commented:
I addressed the previous issues and added some tests. I dont think the coveralls complaint about simulation coverage doing down is correct. the coverage did tick up about 5% with this.
